src/PluginCreator.ts

```diff
@@ -1,209 +1,208 @@
-import * as ts from 'typescript'
-import * as resolve from 'resolve'
-import * as path from 'path'
-import compareVersions from 'compare-versions'
+import * as ts from 'typescript';
+import * as resolve from 'resolve';
+import * as path from 'path';
+import compareVersions from 'compare-versions';
+import { parseConfigFileTextToJson } from 'typescript/lib/tsserverlibrary';
 
-export type TransformerPlugin = {
-    before?: ts.TransformerFactory<ts.SourceFile>
-    after?: ts.TransformerFactory<ts.SourceFile>
-    afterDeclaration?: ts.TransformerFactory<ts.SourceFile>
-} | ts.TransformerFactory<ts.SourceFile>
-
-export type FactoryType = 'ls' | 'program' | 'opts' | 'checker'
+export type FactoryType = 'ls' | 'program' | 'opts' | 'checker';
 
 export type PluginConfig = {
-    name?: string
-    transform?: string
-    type?: FactoryType
-    after?: boolean
-    before?: boolean
-    afterDeclaration?: boolean
-    [options: string]: any;
-}
+    name?: string;
+    transform?: string;
+    type: FactoryType;
+    after?: boolean;
+    before?: boolean;
+    afterDeclaration?: boolean;
+};
 
-export type PluginFactory = {
-    type: 'ls'
-    (ls: ts.LanguageService, options?: PluginConfig): TransformerPlugin
-} | {
-    type?: 'program'
-    (program: ts.Program, options?: PluginConfig): TransformerPlugin
-} | {
-    type: 'opts'
-    (opts: ts.CompilerOptions, options?: PluginConfig): TransformerPlugin
-} | {
-    type: 'checker'
-    (opts: ts.TypeChecker, options?: PluginConfig): TransformerPlugin
-}
-
-export type TransformerHost = ts.LanguageService | ts.Program | ts.CompilerOptions | ts.TypeChecker
+export type TsTransformerFactory = ts.TransformerFactory<ts.SourceFile>;
 
-export type DeclarationPatchBaseHost = Pick<typeof ts, 'versionMajorMinor'>
+export type TransformerPlugin = {
+    before?: TsTransformerFactory;
+    after?: TsTransformerFactory;
+    afterDeclaration?: TsTransformerFactory;
+};
+
+export type FactoryRet = TransformerPlugin | TsTransformerFactory;
+export type LSPattern = (ls: ts.LanguageService, config?: PluginConfig) => FactoryRet;
+export type ProgramPattern = (program: ts.Program, config?: PluginConfig) => FactoryRet;
+export type CompilerOptionsPattern = (opts: ts.CompilerOptions, config?: PluginConfig) => FactoryRet;
+export type TypeCheckerPattern = (opts: ts.TypeChecker, config?: PluginConfig) => FactoryRet;
+export type DefaultPattern = (context: ts.TransformationContext, program?: ts.Program) => ts.Transformer<ts.SourceFile>;
```

Comment:
Why not ProgramPattern is default?
How much transforms use ``` context: ts.TransformationContext, program?: ts.Program ``` signature?

Reply:
But really this interface not used in transformers without some wrappers:

https://github.com/firede/ts-transform-graphql-tag/blob/master/src/transformer.ts
```ts
export default function(): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
    return (sourceFile: ts.SourceFile) => ts.visitNode(sourceFile, getVisitor(context))
  }
}
```

https://github.com/kimamula/ts-transformer-keys/blob/master/transformer.ts
```ts
export default function transformer(program: ts.Program): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext) => (file: ts.SourceFile) => visitNodeAndChildren(file, program, context);
}
```

https://github.com/longlho/ts-transform-css-modules/blob/master/src/transform.ts#L106
```ts
export default function(opts: Opts) {
    require('css-modules-require-hook')(opts)
    return (ctx: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
        return (sf: ts.SourceFile) => ts.visitNode(sf, visitor(ctx, sf))
    }
}
```

Reply:
Ok, you're right, but then we need also to add new pattern for transformers without factory wrapper:
 `raw`
`opts` -> plugin config 
`compilerOptions` -> compileroptions 

https://github.com/dsherret/ts-nameof/blob/8dec4b18c10c95460ef4c55b154a83aa792bcb06/src/transformation/transformerFactory.ts#L7
```ts
const transformerFactory: ts.TransformerFactory<ts.SourceFile> = context => {
    return file => visitSourceFile(file, context) as ts.SourceFile;
};
```

Reply:
Yes, need separate adapter for transformers without factory. But my question was about default common pattern. Is ts-nameof example so common pattern to be default?

Reply:
Fix default to program pattern